src/main/java/org/opensearch/knn/index/query/filtered/NestedFilteredKNNIterator.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.query.filtered;
+
+import org.apache.lucene.index.BinaryDocValues;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.util.BitSet;
+import org.opensearch.knn.index.SpaceType;
+
+import java.io.IOException;
+
+/**
+ * This iterator iterates filterIdsArray to score. However, it dedupe docs per each parent doc
+ * of which ID is set in parentBitSet and only return best child doc with the highest score.
+ */
+public class NestedFilteredKNNIterator<T extends KNNQueryVector> extends FilteredKNNIterator<T> {
+    private final BitSet parentBitSet;
+    private int currentParent = -1;
+    private int bestChild = -1;
+
+    public NestedFilteredKNNIterator(
+        final int[] filterIdsArray,
+        final T queryVector,
+        final BinaryDocValues values,
+        final SpaceType spaceType,
+        final BitSet parentBitSet
+    ) {
+        super(filterIdsArray, queryVector, values, spaceType);
+        this.parentBitSet = parentBitSet;
+    }
+
+    @Override
+    public int nextDoc() throws IOException {
+        if (currentPos >= filterIdsArray.length) {
+            return DocIdSetIterator.NO_MORE_DOCS;
+        }
+        currentScore = Float.NEGATIVE_INFINITY;
+        currentParent = parentBitSet.nextSetBit(filterIdsArray[currentPos]);
+        do {
+            int currentChild = filterIdsArray[currentPos];
+            values.advance(currentChild);
+            final float score = queryVector.score(values, spaceType);
+            if (score > currentScore) {
+                bestChild = currentChild;
+                currentScore = score;
+            }
+            currentPos++;
+        } while (currentPos < filterIdsArray.length && filterIdsArray[currentPos] < currentParent);
+
+        return bestChild;
```

Comment:
nit: Can this be converted into while loop?

Reply:
changed